vra/resource_block_device_snapshot.go

```diff
@@ -0,0 +1,211 @@
+		Pending:    []string{models.RequestTrackerStatusINPROGRESS},
+		Refresh:    BlockDeviceSnapshotStateRefreshFunc(*apiClient, *createDiskSnapshotCreated.Payload.ID),
+		Target:     []string{models.RequestTrackerStatusFINISHED},
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		MinTimeout: 5 * time.Second,
+	}
+
+	_, err = stateChangeFunc.WaitForState()
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Finished to create vra_block_device_snapshot resource with for vra_block_device: %s", id)
+
+	return resourceBlockDeviceSnapshotRead(d, m)
+}
+
+func BlockDeviceSnapshotStateRefreshFunc(apiClient client.MulticloudIaaS, id string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		ret, err := apiClient.Request.GetRequestTracker(request.NewGetRequestTrackerParams().WithID(id))
+		if err != nil {
+			return "", models.RequestTrackerStatusFAILED, err
+		}
+
+		status := ret.Payload.Status
+		switch *status {
+		case models.RequestTrackerStatusFAILED:
+			return []string{""}, *status, fmt.Errorf(ret.Payload.Message)
+		case models.RequestTrackerStatusINPROGRESS:
+			return [...]string{id}, *status, nil
+		case models.RequestTrackerStatusFINISHED:
+			snapshotID := *ret.Payload.ID
+			return snapshotID, *status, nil
+		default:
+			return [...]string{id}, ret.Payload.Message, fmt.Errorf("BlockDeviceSnapshotStateRefreshFunc: unknown status %v", *status)
+		}
+	}
+}
+
+func resourceBlockDeviceSnapshotRead(d *schema.ResourceData, m interface{}) error {
+	blockDeviceID := d.Get("block_device_id").(string)
+	log.Printf("Reading the vra_block_device_snapshot resource for vra_block_device %s ", blockDeviceID)
+	apiClient := m.(*Client).apiClient
+
+	resp, err := apiClient.Disk.GetDiskSnapshots(disk.NewGetDiskSnapshotsParams().WithID(blockDeviceID))
+	// TODO need to change to the correct behavior after the API is updated
+	//resp, err := apiClient.Disk.GetDiskSnapshot(disk.NewGetDiskSnapshotParams().WithID(blockDeviceID).WithId1(d.Id()))
+	if err != nil {
+		switch err.(type) {
+		case *disk.GetDiskSnapshotNotFound:
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	errMsg := "failed to find the created snapshot for the vra_block_device_snapshot resource with id %s"
+
+	diskSnapshots := resp.Payload
+	if len(diskSnapshots) < 1 {
+		return fmt.Errorf(errMsg, d.Get("id"))
+	}
+
+	for _, diskSnapshot := range diskSnapshots {
+		if diskSnapshot.IsCurrent {
+			d.Set("created_at", diskSnapshot.CreatedAt)
+			d.Set("description", diskSnapshot.Desc)
+			d.Set("name", diskSnapshot.Name)
+			d.SetId(*diskSnapshot.ID)
+			d.Set("is_current", diskSnapshot.IsCurrent)
+			d.Set("org_id", diskSnapshot.OrgID)
+			d.Set("owner", diskSnapshot.Owner)
+			d.Set("updated_at", diskSnapshot.UpdatedAt)
+			d.Set("links", flattenLinks(diskSnapshot.Links))
+			log.Printf("Finished reading the vra_block_device_snapshot resource with id %s", *diskSnapshot.ID)
+			return nil
+		}
+	}
+
+	return fmt.Errorf(errMsg, d.Get("id"))
+}
```

Comment:
This is incorrect to do here, as the state changes to the latest snapshot id and attributes each time there is a new snapshot is created.

Run the GET all disk snapshots and look up for the snapshot id by isCurrent, should be done in Create().
Once you find the snapshot with isCurrent=true, set the `id` and in Read(), you can use the `block_device_id` and `id` attributes to invoke the GET disk snapshot API directly.

Reply:
Thanks, fixed it.